api/provider/workos.go

```diff
@@ -0,0 +1,122 @@
+	FirstName      string                 `mapstructure:"first_name"`
+	LastName       string                 `mapstructure:"last_name"`
+	Object         string                 `mapstructure:"object"`
+	IdpID          string                 `mapstructure:"idp_id"`
+	RawAttributes  map[string]interface{} `mapstructure:"raw_attributes"`
+}
+
+// NewWorkOSProvider creates a WorkOS account provider.
+func NewWorkOSProvider(ext conf.OAuthProviderConfiguration, query *url.Values) (OAuthProvider, error) {
+	if err := ext.Validate(); err != nil {
+		return nil, err
+	}
+	apiPath := chooseHost(ext.URL, defaultWorkOSAPIBase)
+
+	// Attach custom query parameters to the WorkOS authorization URL.
+	// See https://workos.com/docs/reference/sso/authorize/get.
+	var authCodeOptions []oauth2.AuthCodeOption
+	if query != nil {
+		if connection := query.Get("connection"); connection != "" {
+			authCodeOptions = append(authCodeOptions, oauth2.SetAuthURLParam("connection", connection))
+		} else if organization := query.Get("organization"); organization != "" {
+			authCodeOptions = append(authCodeOptions, oauth2.SetAuthURLParam("organization", organization))
+		} else if provider := query.Get("workos_provider"); provider != "" {
+			authCodeOptions = append(authCodeOptions, oauth2.SetAuthURLParam("provider", provider))
+		}
+
+		if login_hint := query.Get("login_hint"); login_hint != "" {
+			authCodeOptions = append(authCodeOptions, oauth2.SetAuthURLParam("login_hint", login_hint))
+		}
+	}
+
+	return &workosProvider{
+		Config: &oauth2.Config{
+			ClientID:     ext.ClientID,
+			ClientSecret: ext.Secret,
+			Endpoint: oauth2.Endpoint{
+				AuthURL:  apiPath + "/sso/authorize",
+				TokenURL: apiPath + "/sso/token",
+			},
+			RedirectURL: ext.RedirectURI,
+		},
+		APIPath:         apiPath,
+		AuthCodeOptions: authCodeOptions,
+	}, nil
+}
+
+func (g workosProvider) AuthCodeURL(state string, args ...oauth2.AuthCodeOption) string {
+	opts := append(args, g.AuthCodeOptions...)
+	return g.Config.AuthCodeURL(state, opts...)
+}
+
+func (g workosProvider) GetOAuthToken(code string) (*oauth2.Token, error) {
+	return g.Exchange(oauth2.NoContext, code)
+}
+
+func (g workosProvider) GetUserData(ctx context.Context, tok *oauth2.Token) (*UserProvidedData, error) {
+	if tok.AccessToken == "" {
+		return &UserProvidedData{}, nil
+	}
+
+	// WorkOS API returns the user's profile data along with the OAuth2 token, so
+	// we can just convert from `map[string]interface{}` to `workosUser` without
+	// an additional network request.
+	var u workosUser
+	err := mapstructure.Decode(tok.Extra("profile"), &u)
+	if err != nil {
+		return nil, err
+	}
+
+	if u.Email == "" {
+		return nil, errors.New("Unable to find email with WorkOS provider")
+	}
+
+	return &UserProvidedData{
+		Metadata: &Claims{
+			Issuer:        g.APIPath,
+			Subject:       u.ID,
+			Name:          strings.TrimSpace(u.FirstName + " " + u.LastName),
+			Email:         u.Email,
+			EmailVerified: true,
```

Comment:
i assume emails returned by workos are guaranteed to be verified?

Reply:
Actually they aren't: should I delete that line? I saw that some other providers had that too, even if their providers didn't guarantee so, so I did the same. I couldn't figure out the importance of that though: what does this affect?

Reply:
Right, confirmation emails are disabled for us so this wasn't on my mind. I can remove it if necessary, although should that be done for the other providers as well?

Reply:
other providers don't return the email at all if the email is not verified, do you happen to know if this is the case for workos as well?

Reply:
I believe WorkOS returns the email based on what the SAML SSO it's integrated with returns, without verifying if the email actually exists. However, I don't think this is much of an issue, since users have to manually add these SAML SSO integrations into WorkOS themselves, and if they do so it means that they trust these SAML SSO providers on the other end to return the correct results.

I think it might make sense to drop the `EmailVerified: true` though. WDYT?

Reply:
Ah i see, then in that case, i actually think it's fine for `EmailVerified: true`. The [original SAML implementation](https://github.com/supabase/gotrue/blob/master/api/external_saml.go#L53) by netlify also assumes that the email used is verified.

Reply:
:+1:, I've changed back to that behaviour.